### ISSUE TITLE
Tuning parameters to SamAutomaticMaskGenerator and IndexError: too many indices for tensor of dimension 1

### ISSUE THREAD

**Issue by chenjia-lin** [facebookresearch/segment-anything#614]
I'm using the automatic mask generator to segment some images and I got the following error a couple times.
```
Cell In[83], line 1
----> 1 sample_mask = fin_mask_generator.generate(sample_img)

File ~\AppData\Roaming\Python\Python311\site-packages\torch\utils\_contextlib.py:115, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    112 @functools.wraps(func)
    113 def decorate_context(*args, **kwargs):
    114     with ctx_factory():
--> 115         return func(*args, **kwargs)

File ~\AppData\Roaming\Python\Python311\site-packages\segment_anything\automatic_mask_generator.py:163, in SamAutomaticMaskGenerator.generate(self, image)
    138 """
    139 Generates masks for the given image.
    140 
   (...)
    159          the mask, given in XYWH format.
    160 """
    162 # Generate masks
--> 163 mask_data = self._generate_masks(image)
    165 # Filter small disconnected regions and holes in masks
    166 if self.min_mask_region_area > 0:

File ~\AppData\Roaming\Python\Python311\site-packages\segment_anything\automatic_mask_generator.py:212, in SamAutomaticMaskGenerator._generate_masks(self, image)
    209 # Remove duplicate masks between crops
    210 if len(crop_boxes) > 1:
    211     # Prefer masks from smaller crops
--> 212     scores = 1 / box_area(data["crop_boxes"])
    213     scores = scores.to(data["boxes"].device)
    214     keep_by_nms = batched_nms(
    215         data["boxes"].float(),
    216         scores,
    217         torch.zeros_like(data["boxes"][:, 0]),  # categories
    218         iou_threshold=self.crop_nms_thresh,
    219     )

File ~\AppData\Roaming\Python\Python311\site-packages\torchvision\ops\boxes.py:235, in box_area(boxes)
    233     _log_api_usage_once(box_area)
    234 boxes = _upcast(boxes)
--> 235 return (boxes[:, 2] - boxes[:, 0]) * (boxes[:, 3] - boxes[:, 1])

IndexError: too many indices for tensor of dimension 1
```
For context, I am trying to get a single mask that covers either the interior or exterior of the object in the image below, and I have 138 other images which look almost identical to it.
<img src="https://github.com/facebookresearch/segment-anything/assets/88413061/3a1f311d-215f-40b0-ab88-5ee7b989976f" height="300"/>
I used the following instance of `SamAutomaticMaskGenerator` and noticed that it worked for five images but failed for 134 others (all of which gave an `IndexError`).
```
SamAutomaticMaskGenerator(
    model=sam,
    points_per_side=8,
    pred_iou_thresh=0.88,
    stability_score_thresh=0.99,
    crop_n_layers=1,
    crop_n_points_downscale_factor=2,
    min_mask_region_area=1000, 
)
```
Tweaking the parameters somewhat fixed the issue (e.g. changing `stability_score_thresh` to 0.97), but I want to avoid doing this extensively for multiple sets of 100+ images. I'm not familiar with Torch, so I couldn't figure out why the generator worked for some images but not others (despite them being very similar). I also wanted to know if someone has suggestions for which parameters I should tweak for my particular task.

**Comment by heyoeyo** [comment]
Running the image you provided (and the config for the mask generator), what ends up happening is that `data["crop_boxes"]` is an empty array, so it has a shape of [0], which doesn't match the expected Nx4 shape the rest of the code is assuming, leading to the error.

I don't know much about how the cropping works, but I guess this implies that the mask generator isn't able to find any masks? It's possible to fix the error by only running the [scoring/filtering code](https://github.com/facebookresearch/segment-anything/blob/6fdee8f2727f4506cfbbe553e23b895e27956588/segment_anything/automatic_mask_generator.py#L212C13-L220C37) when there is actually data to work with:
```python
if data["crop_boxes"].numel() > 0:
    scores = 1 / box_area(data["crop_boxes"])
    scores = scores.to(data["boxes"].device)
    keep_by_nms = batched_nms(
        data["boxes"].float(),
        scores,
        torch.zeros_like(data["boxes"][:, 0]),  # categories
        iou_threshold=self.crop_nms_thresh,
    )
    data.filter(keep_by_nms)
```
Basically just putting all of the scoring/filtering code under an if statement that makes sure there is at least one element in `data["crop_boxes"]`.

However, since there is no data to store, the outcome from this (at least on the image you provided) is nothing! No masks. So I'm not sure that this is any better than fiddling with parameters unfortunately.

**Comment by chenjia-lin** [comment]
Thanks for the clarification! I suspected that the generator probably had some trouble getting a mask with specs that fell within the thresholds, but I didn't know how to test it. I might very well need to mess with the parameters until I get the desired result for each set of images.
